### PR TITLE
tree: change the fold function to be more expressive

### PR DESCRIPTION
**Pull request by samoht**

### CODE REVIEW

**Comment by samoht** [comment]
This is now ready for review.

File: src/irmin/tree_intf.ml [hunk 1]
```diff
@@ -184,12 +185,19 @@ module type S = sig
   type 'a node_fn = key -> step list -> 'a -> 'a Lwt.t
   (** The type for {!fold}'s [pre] and [post] parameters. *)
 
+  type depth = [ `Eq of int | `Le of int | `Lt of int ] [@@deriving irmin]
```
Comment (CraigFe) [inline src/irmin/tree_intf.ml:188]:
```suggestion
  type depth = [ `Eq of int | `Lt of int ] [@@deriving irmin]
```
Seems cleaner to have express the `Le` case in terms of `Lt` or vice versa.

**Review by CraigFe — APPROVED** [review]

File: src/irmin/tree.ml [hunk 2]
```diff
@@ -690,53 +693,73 @@ module Make (P : S.PRIVATE) = struct
...
+              to_map t >>= function
+              | Ok m ->
+                  if force = `And_clear then clear ~depth:0 t;
+                  (map [@tailcall]) ~path acc d (Some m) k
+              | Error (`Dangling_hash _) -> (map [@tailcall]) ~path acc d None k
+              )
+          | `False skip -> (
+              match t.info.map with
+              | Some n -> (map [@tailcall]) ~path acc d (Some n) k
+              | _ -> skip path acc)
```
Comment (Ngoguey42) [inline src/irmin/tree.ml:717]: Hi there :)

Shouldn't `k` be called after skip?

Comment (samoht) [reply src/irmin/tree.ml:717]: Hey! Yes good catch, I'll fix it in another PR :-)

**Comment by samoht** [comment]
@CraigFe I've added even more cases to that `depth` datatype. Let me know what you think ;-)

**Review by CraigFe — APPROVED** [review]
Oddly, I'm happier with it this time around :wink:

File: src/irmin/tree.ml (hunk 2 above)
Comment (samoht) [reply src/irmin/tree.ml:717]: See #781

**Pull request merged**